File: src/creations/news_desk/imports.py
"""News-desk import provider (sidecar side).

The user-facing "import subtitle / import chapters" flows for the news_desk
workbench. Both pull from the bound material and SNAPSHOT into this creation
instance ([[project_snapshot_principle]]): a subtitle component copies the
chosen language's SRT into the instance dir and points its srt_path at it; the
chapter component copies the analysis.json chapter rows into its schedule.

Registered on CreationType.import_provider so core_rpc resolves it generically
(ADR-0004 — the base layer never imports this by name). This round covers
importing from the bound material's existing artifacts only (the common case
after ASR / chapter analysis has run); external file import is a follow-up.

  list_imports(project, instance)                              -> {subtitleLangs, analyses}
  import_resource(project, instance, component_id, params)     -> updated component dict

`params` is provider-defined and opaque to the base layer:
  {"kind": "subtitle", "lang": "<iso>"}           — snapshot that language's SRT
  {"kind": "chapters", "filename": "<analysis>"}  — fill schedule from analysis
"""

from __future__ import annotations

import os
import shutil
from typing import Any

from creations.news_desk.config import NewsDeskInstanceConfig


def _resolve(project, instance: str):
    """(cfg, inst_dir, model | None). model is None when unbound. Mirrors
    export.py::_resolve — resolves the bound material via the registry without
    hard-coding a plugin name."""
    inst_dir = project.creation_instance_dir("news_desk", instance)
    cfg = NewsDeskInstanceConfig.load(os.path.join(inst_dir, "config.json"))
    if cfg.bound_material is None:
        return cfg, inst_dir, None

    import materials

    mtype = materials.get(cfg.bound_material.type_name)
    model = (
        mtype.instance_factory(project, cfg.bound_material.instance_name)
        if mtype and mtype.instance_factory
        else None
    )
    return cfg, inst_dir, model


def list_imports(project, instance: str) -> dict[str, Any]:
    """What this instance can import from its bound material: subtitle languages
    (for subtitle components) and analysis filenames (for the chapter component).
    Empty lists when unbound or the model lacks the artifacts."""
    _cfg, _inst_dir, model = _resolve(project, instance)
    if model is None:
        return {"subtitleLangs": [], "analyses": []}
    langs = model.list_subtitle_languages() if hasattr(model, "list_subtitle_languages") else []
    analyses = model.list_analyses() if hasattr(model, "list_analyses") else []
    return {"subtitleLangs": list(langs), "analyses": list(analyses)}


def _find_component(cfg: NewsDeskInstanceConfig, component_id: str) -> dict:
    for c in cfg.components:
        if c.get("id") == component_id:
            return c
    raise ValueError(f"no component with id {component_id!r}")


def _import_subtitle(inst_dir, cfg, model, component, lang: str) -> dict:
    """Snapshot the bound material's <lang>.srt into the instance dir and point
    the subtitle component's srt_path at it (faithful to the Tk _import_srt
    snapshot, but sourcing from the material instead of a file dialog)."""
    if component.get("kind") != "subtitle":
        raise ValueError("import subtitle: component is not a subtitle")
    src = model.subtitle_path(lang)
    if not os.path.isfile(src):
        raise ValueError(f"subtitle not found for language {lang!r}")
    rel = f"subtitles/{component['id']}.srt"
    dst = os.path.join(inst_dir, rel)
    os.makedirs(os.path.dirname(dst), exist_ok=True)
    shutil.copyfile(src, dst)
    component["srt_path"] = rel
    cfg.save(os.path.join(inst_dir, "config.json"))
    return component


def _import_chapters(inst_dir, cfg, model, component, filename: str) -> dict:
    """Fill the chapter component's schedule from an analysis.json envelope
    (faithful to the Tk _import_from_analysis snapshot)."""
    if component.get("kind") != "chapter":
        raise ValueError("import chapters: component is not a chapter")
    try:
        env = model.read_analysis(filename)
    except (OSError, ValueError) as exc:
        raise ValueError(f"read analysis failed: {exc}") from exc
    chapters = env.get("chapters") if isinstance(env, dict) else None
    if not chapters:
        raise ValueError("analysis has no chapters")
    schedule = [
        {
            "start_sec": float(ch.get("start_sec", 0.0)),
            "end_sec": float(ch.get("end_sec", 0.0)),
            "title": str(ch.get("title", "")),
            "refined": str(ch.get("refined", "")),
            "key_points": list(ch.get("key_points") or []),
        }
        for ch in chapters
        if isinstance(ch, dict)
    ]
    component["schedule"] = schedule
    cfg.save(os.path.join(inst_dir, "config.json"))
    return component


def import_resource(project, instance: str, component_id: str, params: dict) -> dict:
    """Perform an import into one component and return the updated component dict.
    The single owner persists the mutation (cfg.save). See module docstring for
    the params shape."""
    if not isinstance(params, dict):
        raise ValueError("import params must be an object")
    cfg, inst_dir, model = _resolve(project, instance)
    if model is None:
        raise ValueError("creation is not bound to a material")
    component = _find_component(cfg, component_id)
    kind = params.get("kind")
    if kind == "subtitle":
        return _import_subtitle(inst_dir, cfg, model, component, str(params.get("lang", "")))
    if kind == "chapters":
        return _import_chapters(inst_dir, cfg, model, component, str(params.get("filename", "")))
    raise ValueError(f"unknown import kind: {kind!r}")

File: desktop/src/creations/news_desk/imports.ts
/**
 * News-desk import logic (ADR-0008 B4 TS port of `src/creations/news_desk/imports.py`).
 *
 * The "import subtitle / import chapters" flows: both pull from the bound material
 * and SNAPSHOT into this creation instance ([[project_snapshot_principle]]) — a
 * subtitle component copies the chosen language's SRT into the instance dir and
 * points its srt_path at it; a chapter component copies the analysis.json chapter
 * rows into its schedule. The single owner persists the mutation (owner.save()).
 *
 * Pure + injectable (owner + Fs + a minimal material accessor) so vitest drives it
 * with an in-memory Fs; bound-material resolution stays in clientBackend (RPC).
 *
 *   params {kind:"subtitle", lang}            — snapshot that language's SRT
 *   params {kind:"chapters", filename}        — fill schedule from analysis.json
 */

import type { Fs } from "../../renderer/ipc/fs";
import type { ComponentDict } from "./componentDefs";
import type { NewsDeskConfigOwner } from "./configOwner";

/** One importable dubbing version (a voice under a language). */
export interface DubImport {
  lang: string;
  id: number;
  name: string;
}

/** One dub version with its resolved source audio (model.dubVersions shape). */
interface DubVersionLite {
  id: number;
  name: string;
  audioPath: string;
}

/** The subset of NewsVideoModel imports needs (structural — lets vitest fake it). */
export interface ImportMaterial {
  listSubtitleLanguages(): Promise<string[]>;
  listAnalyses(): Promise<string[]>;
  subtitlePath(lang: string): string;
  readAnalysis(filename: string): Promise<Record<string, unknown>>;
  /** Synthesized dubbing versions (voices) for a language. */
  dubVersions(lang: string): Promise<DubVersionLite[]>;
}

export async function listNewsDeskImports(
  model: ImportMaterial,
): Promise<{ subtitleLangs: string[]; analyses: string[]; dubVersions: DubImport[] }> {
  const subtitleLangs = await model.listSubtitleLanguages();
  // Every dub version across all languages — one importable entry per voice.
  const dubVersions: DubImport[] = [];
  for (const lang of subtitleLangs) {
    try {
      for (const v of await model.dubVersions(lang)) dubVersions.push({ lang, id: v.id, name: v.name });
    } catch {
      /* skip a language whose dub manifest can't be read */
    }
  }
  return { subtitleLangs, analyses: await model.listAnalyses(), dubVersions };
}

/** Perform one import into a component and return the updated component dict.
 *  Mutates the owner in place and persists via owner.save(); `fs` is used for the
 *  SRT copy (same backing store the owner saves through). */
export async function importNewsDeskResource(
  owner: NewsDeskConfigOwner,
  fs: Fs,
  instanceDir: string,
  model: ImportMaterial,
  componentId: string,
  params: Record<string, unknown>,
): Promise<ComponentDict> {
  if (!params || typeof params !== "object") throw new Error("import params must be an object");
  const comp = owner.components.find((c) => c["id"] === componentId);
  if (!comp) throw new Error(`no component with id ${componentId}`);
  const kind = params["kind"];

  if (kind === "subtitle") {
    if (comp["kind"] !== "subtitle") throw new Error("import subtitle: component is not a subtitle");
    const lang = String(params["lang"] ?? "");
    const src = model.subtitlePath(lang);
    if (!(await fs.stat(src)).exists) throw new Error(`subtitle not found for language ${lang}`);
    const rel = `subtitles/${componentId}.srt`;
    await fs.copy(src, `${instanceDir}/${rel}`);
    comp["srt_path"] = rel;
    await owner.save();
    return comp;
  }

  if (kind === "dubbing") {
    if (comp["kind"] !== "dubbing") throw new Error("import dubbing: component is not a dubbing track");
    const lang = String(params["lang"] ?? "");
    const versionId = Number(params["version_id"] ?? -1);
    const v = (await model.dubVersions(lang)).find((x) => x.id === versionId);
    if (!v) throw new Error(`dubbing version not found: ${lang} #${versionId}`);
    if (!(await fs.stat(v.audioPath)).exists) throw new Error(`dubbing audio missing: ${v.audioPath}`);
    const rel = `audio/${componentId}.mp3`;
    await fs.copy(v.audioPath, `${instanceDir}/${rel}`);
    comp["audio_path"] = rel;
    await owner.save();
    return comp;
  }

  if (kind === "chapters") {
    if (comp["kind"] !== "chapter") throw new Error("import chapters: component is not a chapter");
    const filename = String(params["filename"] ?? "");
    let env: Record<string, unknown>;
    try {
      env = await model.readAnalysis(filename);
    } catch (e) {
      throw new Error(`read analysis failed: ${e instanceof Error ? e.message : String(e)}`);
    }
    const chapters = Array.isArray(env["chapters"]) ? (env["chapters"] as unknown[]) : null;
    if (!chapters || chapters.length === 0) throw new Error("analysis has no chapters");
    comp["schedule"] = chapters
      .filter((ch): ch is Record<string, unknown> => !!ch && typeof ch === "object")
      .map((r) => ({
        start_sec: Number(r["start_sec"] ?? 0) || 0,
        end_sec: Number(r["end_sec"] ?? 0) || 0,
        title: String(r["title"] ?? ""),
        refined: String(r["refined"] ?? ""),
        key_points: Array.isArray(r["key_points"]) ? r["key_points"] : [],
      }));
    // Snapshot the AI-suggested video titles too: the same analysis call emits
    // both chapters and a top-level titles[]. publish.md's "Candidate Titles"
    // section reads them off the chapter component (render.ts), so without this
    // they silently never appear in the export.
    comp["titles"] = Array.isArray(env["titles"])
      ? (env["titles"] as unknown[]).map((t) => String(t).trim()).filter(Boolean)
      : [];
    await owner.save();
    return comp;
  }

  throw new Error(`unknown import kind: ${String(kind)}`);
}
